tests/0101-fetch-from-follower.cpp

```diff
@@ -0,0 +1,320 @@
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "testcpp.h"
+
+#if WITH_RAPIDJSON
+
+#include <iostream>
+#include <cstring>
+#include <cstdlib>
+#include <assert.h>
+#include <sstream>
+#include <string>
+#include <map>
+
+#include <rapidjson/document.h>
+#include <rapidjson/schema.h>
+#include <rapidjson/filereadstream.h>
+#include <rapidjson/stringbuffer.h>
+#include <rapidjson/error/en.h>
+#include <rapidjson/prettywriter.h>
+
+
+/**
+ * @brief A basic test of fetch from follower funtionality
+ *  - produces a bunch of messages to a replicated topic.
+ *  - configure the consumer such that `client.rack` is different from the
+ *    broker's `broker.rack` (and use
+ *    org.apache.kafka.common.replica.RackAwareReplicaSelector).
+ *  - consume the messages, and check they are as expected.
+ *  - use rxbytes from the statistics event to confirm that 
+ *    the messages were retrieved from the replica broker (not the
+ *    leader).
+ */
+
+
+static void test_assert(bool cond, std::string msg) {
+  if (!cond)
+    Test::Say(msg);
+  assert(cond);
+}
+
+
+class TestEvent2Cb : public RdKafka::EventCb {
+ public:
+  static bool should_capture_stats;
+  static bool has_captured_stats;
+  static std::map<int32_t, int64_t> rxbytes;
+
+  void event_cb (RdKafka::Event &event) {
+
+    switch (event.type())
+    {
+      case RdKafka::Event::EVENT_LOG:
+        Test::Say(event.str() + "\n");
+        break;
+      case RdKafka::Event::EVENT_STATS:
+        if (should_capture_stats) {
+
+          rapidjson::Document d;
+          if (d.Parse(event.str().c_str()).HasParseError())
+            Test::Fail(tostr() << "Failed to parse stats JSON: " <<
+                       rapidjson::GetParseError_En(d.GetParseError()) <<
+                       " at " << d.GetErrorOffset());
+
+          /* iterate over brokers. */
+          rapidjson::Pointer jpath((const char *)"/brokers");
+          rapidjson::Value *pp = rapidjson::GetValueByPointer(d, jpath);
+          if (pp == NULL)
+            return;
+
+          for (rapidjson::Value::ConstMemberIterator itr = pp->MemberBegin(); itr != pp->MemberEnd(); ++itr) {
```

Comment:
C++ is so noisy and verbose

Reply:
IIRC C++26 is slated to sort this out with additional languages features and a rewrite of the standard library